### ISSUE TITLE
Would you accept a PR that converts this project to build with TypeScript?

### ISSUE THREAD

**Issue by octogonz** [prettier/prettier#6286]
I've been studying the Prettier code and considering contributing some fixes.

However, I am finding the source code to be rather difficult to follow, because there are no type declarations anywhere.  For example, here's a typical function:

```js
function isWithinParentArrayProperty(path, propertyName) {
  const node = path.getValue();
  const parent = path.getParentNode();

  if (parent == null) {
    return false;
  }

  if (!Array.isArray(parent[propertyName])) {
    return false;
  }

  const key = path.getName();
  return parent[propertyName][key] === node;
}
```

For a newcomer, it takes some effort to figure out:
- What kind of thing is `path`?
- What kind of thing is `parent`?
- Which of these things is allowed to be `null` (or `undefined`)?

The AST data structure in particular is fairly complicated.  I recently worked on another project that used [Recast](https://github.com/benjamn/recast), and there the TypeScript validation caught a lot of mistakes.  At every step of the AST tree, the type information clearly specifies what kind of nodes are allowable, and the IntelliSense shows you the available properties.

Converting the Prettier code base to TypeScript is obviously a nontrivial amount of work.  But if someone from the community volunteered to do this, would you accept such a PR?

**Comment by evilebottnawi** [comment]
I don't think we need typescript here, just my opinion, i love typescript, but not here

**Comment by duailibe** [comment]
Thanks for the offer! We have considered this path in the past, but one restriction we still value is being able to install Prettier directly from GitHub: this enables our users to be unblocked in case they are affected with a bug, or need a feature, until we release a new version.

It also lowers the barrier for people to fork Prettier. In that case, they don't need to publish a new package for that.

This restriction makes it hard to use TypeScript (or any other "preprocessor").

**Comment by octogonz** [comment]
> This restriction makes it hard to use TypeScript (or any other "preprocessor").

@duailibe Wouldn't these same considerations apply generally to any project that uses any sort of transpiler (Jest, Yarn, Angular, etc.)?  Those projects seem to have concluded that readability and validation are important enough concerns to be worth the extra overhead.

> It also lowers the barrier for people to fork Prettier. In that case, they don't need to publish a new package to use their fork in their projects.

This could be solved by committing the build output, and then pushing it to a GitHub release branch, right?

**Comment by dasa** [comment]
Maybe using TypeScript on the js source code like this would be useful: https://devblogs.microsoft.com/typescript/how-to-upgrade-to-typescript-without-anybody-noticing-part-1/

**Comment by evilebottnawi** [comment]
@dasa yep, using comments for typescript is good idea

**Comment by duailibe** [comment]
@octogonz 

> Wouldn't these same considerations apply generally to any project that uses any sort of transpiler (Jest, Yarn, Angular, etc.)?

Yes, so I'd guess they have a different set of trade-offs.

> This could be solved by committing the build output, and then pushing it to a GitHub release branch, right?

I think so.

@dasa Yes, that's an option. I opened a PR doing that with Flow comments, but I gave up in the end. Nobody did the same with JSDoc yet.
I'm also under the impression that this is less powerful than using TS syntax, so I thought it might be less useful. Didn't investigate it. Happy to learn more about it.

**Comment by thorn0** [comment]
> I'm also under the impression that this is less powerful than using TS syntax, so I thought it might be less useful. Didn't investigate it. Happy to learn more about it.

We've [already discussed](https://github.com/prettier/prettier/pull/4797#issuecomment-403814252) this concern in that PR.

**Comment by octogonz** [comment]
Such gymnastics may not really be necessary.  Let's say that converting all the unit tests is relatively unimportant, and could be done later, incrementally.  The subtree that matters is [prettier/src](https://github.com/prettier/prettier/tree/master/src), since it contains the public APIs and important formatting logic.

I counted 110 .js files in that folder.  If the Prettier project wants to move to TypeScript, we could throw humans at the problem and convert the files manually. Many projects have done this sort of conversion before.  It's well understood and can go pretty fast.  The work is rather fun. 110 files is small enough to complete in a matter of weeks. The conversion could be validated by diffing the TypeScript compiler output against the original sources (since the initial conversion would aim to add type annotations without altering semantics in any way).

So instead of "how would we accomplish this?" the main question is:  "If someone from the community volunteered to do this, would such a PR be accepted?"

**Comment by Shinigami92** [comment]
I'm one of these humans 🙋‍♂️

The current "problem" is that prettier can also be used directly as a dependency from the github repo. 
So if I unterstand correctly some depends on the src folder?!

**Comment by thorn0** [comment]
> Many projects have done this sort of conversion before. It's well understood and can go pretty fast.

It's not always easy '[to capture highly dynamic JS behavior in static types](https://twitter.com/acemarke/status/1149894702415716352)'. Prettier with all the different AST formats it works with seems like one of those projects with highly dynamic JS behavior. TS might work for it, but the conversion might not be exactly pretty fast. I wish I was wrong.

**Comment by evilebottnawi** [comment]
Unfortunately, you are right

**Comment by Skillz4Killz** [comment]
> Thanks for the offer! We have considered it in the past, but one restriction we value is being able to install Prettier directly from GitHub: this enables our users to be unblocked in case they are affected with a bug, or need a feature, until we release a new version.
> 
> It also lowers the barrier for people to fork Prettier. In that case, they don't need to publish a new package to use their fork in their projects.
> 
> This restriction makes it hard to use TypeScript (or any other "preprocessor").

Add a "prepack" script in package.json that does tsc so users can use Typescript repos as well.

**Comment by Shinigami92** [comment]
> Add a "prepack" script in package.json that does tsc so users can use the Github directly or even fork and use their github even if it is a Typescript repo.

Some questions:
- Does someone need TypeScript in their own `package.json` for this to work?
- According to the documentation https://docs.npmjs.com/misc/scripts
   Will this run on the consumer side? Or on pack/publish to npm? Or both?
- What could be a sideeffect?
- So I'm a little bit afraid of installing/running something on consumer side on `install` 😬
- ... Could `typescript` be a peerDependency in `prettier`?

**Comment by octogonz** [comment]
Executing the source code seems like a strange requirement to impose. Other TypeScript/Babel projects don't worry about this. Building code is a standard part of professional software development. If our goal is to make it easier for people to improve Prettier, then TypeScript's improved validation and readability seem like a more important requirement.

If a person is smart enough to modify a complex tool like Prettier, it's probably reasonable to assume they will know how to (1) publish to a private NPM registry or (2) commit their compiled output to a Git branch somewhere.

**Comment by lydell** [comment]
Maybe we could automatically publish alpha versions of Prettier on each push to `master` to avoid the whole “install from github compatible” things 🤷‍♂

**Comment by waynevanson** [comment]
A strategy that might work would be to partially adopt typescript to the build.

If this project builds correctly javascript to javascript with the typescript compiler. There are so many tests in here that errors would be caught before releases.

For the dynamic language feature concerns: overloads, generics and mappable types allow flexible generation of types. I'd be keen to help out on something like this.

**Comment by Shinigami92** [comment]
The last time I looked at adding JSDoc typings, it was incredibly difficult to find some types, especially for large complex objects like `Doc`
This could even be impossible to type with JSDoc...

I would appreciate to rewrite prettier in TypeScript, but there are statements like "This repo can be added as a dependency directly from the master branch" and so on (_I've never used such black magic myself_)

I don't know the current opinion

**Comment by waynevanson** [comment]
@Shinigami92 Someone suggested that perhaps that the master branch could be the built branch, with a different branch as the development branch. That could remove the black magic headache.

**Comment by Shinigami92** [comment]
I would be happy with that. So the master branch would be the builded endresult and the development branch uses github actions to automatically build and push into master?

So master would be something like gh-pages branch?! ^^

**Comment by waynevanson** [comment]
@Shinigami92 Yes, that's correct.

I see there's a `tsconfig.json` in master. Looks like someone's doing something with TS.

Once project is converted to BUILD with TS, I'll start contributing.

**Comment by Shinigami92** [comment]
One of the main TS leads and contributer of this `tsconfig.json`... It's me  
The current state was to first introduce typings with JSDoc  
<img src="https://user-images.githubusercontent.com/7195563/82180682-55a28800-98e1-11ea-8206-f5e4ac12c4f5.png" width="300px">

**Comment by octogonz** [comment]
The original intent of this issue was to decide whether Prettier should join the rest of the popular tools that have moved to TypeScript.  (How best to do it is a separate conversation.)

Since it's been it a year with no consensus, let's change the question:

What is the way to reach a yes/no decision regarding TypeScript?  A community vote with a specified end date?  An authoritative statement from one of the maintainers?

**Comment by JounQin** [comment]
Can we reconsider this in year 2024? 🤣

**Comment by Shinigami92** [comment]
@fisker I think this issue got closed by accident
